### ISSUE TITLE
[AHM] Remove dependency to full exposure for validator identification on RC pallets

### ISSUE THREAD

**Issue by Ank4n** [paritytech/polkadot-sdk#6344]
Currently, we’re tracking historical sessions using the [FullIdentification](https://github.com/polkadot-fellows/runtimes/blob/main/relay/polkadot/src/lib.rs#L479) of a validator (i.e. tuple of validator id and its exposure). Essentially, for each session, we take the validator ID and its exposure, create a trie, and store the root in the state for the past 28 eras.

Then for [equivocation proofs](https://paritytech.github.io/polkadot-sdk/master/pallet_babe/pallet/dispatchables/fn.report_equivocation.html), the validator provides two inputs: the equivocation proof and the key ownership proof. For the latter, the validator generates a proof via a runtime API exposed by historical session pallet, submitting it in the extrinsic to report equivocation. This proof is verified on-chain against the historical root we are tracking.

This will be a problem once staking modules are migrated to AH. The exposure data will reside in AH, and RC should ideally only have the active validator keys. Offence report still has to go to RC to verify equivocation proof, maybe disable validator, and  send the offender (+ session/era) to AH. 

I am not really sure what is the benefit of identifying the validator by its full exposure? I think we can either

- Identify validator only with validator id and not its exposure. We still keep track of historical session roots, but the trie only consists of validator ids.
- Or not need key ownership proof while submitting equivocation proof? Do we really need it? This would imply not use [historical session](https://paritytech.github.io/polkadot-sdk/master/pallet_session/historical/index.html) at all. 

## Update
We decided to go with the first approach
> Identify validator only with validator id and not its exposure. We still keep track of historical session roots, but the trie only consists of validator ids.

**Comment by ordian** [comment]
> I am not really sure what is the benefit of identifying the validator by its full exposure?

My understanding is that it's needed for slashing purposes (offences pallet). If you want to slash someone for an offence that happened 27 days ago due to an equivocation, you need to know their exposure (who nominated them, amounts, etc) along with slash % in order slash accordingly. If all that historical exposure data is stored somewhere else, I don't see a need for it though.

> Or not need key ownership proof while submitting equivocation proof? Do we really need it? 

Equivocations are identified by session keys (e.g. BABE, GRANDPA, BEEFY) and we need a way to map them to validator id (also in the past). This is what historical pallet is for.

We also use historical pallet for dispute slashes: https://github.com/polkadot-fellows/runtimes/blob/3b18d942c766c7f358da50608b44edbe218284a4/relay/polkadot/src/lib.rs#L1410. Because it uses the same offences pallet as a means of slashes.

> Essentially, for each session, we take the validator ID and its exposure, create a trie, and store the root in the state for the past 28 eras.

The amount of data historical pallet stores is already quite lightweight (28 hashes).
If we remove the need to store the exposure and only keep session keys tracking, then relay chain only needs to know about active validator set keys to calculate the historical trie root.
Key ownership proof and slash percent would then be sent to the staking/asset-hub parachain via XCM and that will have to figure out who to slash. Does that sound reasonable?

**Comment by Ank4n** [comment]
> My understanding is that it's needed for slashing purposes (offences pallet). If you want to slash someone for an offence that happened 27 days ago due to an equivocation, you need to know their exposure (who nominated them, amounts, etc) along with slash % in order slash accordingly. If all that historical exposure data is stored somewhere else, I don't see a need for it though.

Yes, historical exposure data is still kept in AH. Just want to get rid of the dependency in RC.

> If we remove the need to store the exposure and only keep session keys tracking, then relay chain only needs to know about active validator set keys to calculate the historical trie root.
Key ownership proof and slash percent would then be sent to the staking/asset-hub parachain via XCM and that will have to figure out who to slash (by probably using a similar mechanism to historical pallet with exposure in it). Does that sound reasonable?

Yes, thank you.

**Comment by burdges** [comment]
> Just want to get rid of the dependency in RC.

Afaik this should be fine.  We need to identify the session in the message from the RC to AH though.

**Comment by Ank4n** [comment]
> If high value contracts can price out PJR checks, then a validator election could be "stolen". We could reserve some AH blocks for PJR tests somehow, or cap overpayment for contract execution, or descide that PJR tests should be moved off of AH sometime in the future.

Most of the election computation are system level, meaning they will happen before any signed extrinsic. The exception is the election solution submitted by the staking miner that we need to prioritize over other extrinsics. Even this has a backup unsigned phase where collators will use OCW to compute the winners (though with reduced election score compared to the signed miner's solution), so I don't see the risk of election being stolen. 

Moving election to another chain is an option for future. However a more feasible alternative would be, if and when elastic scaling becomes possible, we utilise multiple cores during election blocks?

**Comment by burdges** [comment]
> Most of the election computation are system level, meaning they will happen before any signed extrinsic. The exception is the election solution submitted by the staking miner that we need to prioritize over other extrinsics. 

This assumes honest AH nodes, and that AH cannot be censored in other ways, which complicates elastic scaling.  At the parachain retreat, we noticed an honest backers assumption makes this less bad.  We're definitely okay short-term, but this remains messy.

> Even this has a backup unsigned phase where collators will use OCW to compute the winners 

Yeah fair enough, the attack would be:  Your staking miner beats OCW with a highly biases election, after which you censor.  You need to censor for a long time though, so maybe even one honest AH node suffices, excluding the elastic scaling censorship attack.
